src/base/projects/Blob.svelte

```diff
@@ -1,49 +1,59 @@
 <script lang="ts">
   import type { Blob } from "@app/project";
+  import type { MaybeHighlighted } from "@app/syntax";
   import type { ProjectRoute } from "@app/router/definitions";
 
   import HeaderToggleLabel from "@app/base/projects/HeaderToggleLabel.svelte";
-  import ProjectLink from "@app/router/ProjectLink.svelte";
   import Readme from "@app/base/projects/Readme.svelte";
+  import { afterUpdate, onMount } from "svelte";
+  import { highlight } from "@app/syntax";
   import { isMarkdownPath, scrollIntoView, twemoji } from "@app/utils";
-  import { onMount } from "svelte";
+  import { lineNumbersGutter } from "@app/syntax";
+  import { toHtml } from "hast-util-to-html";
   import { updateProjectRoute } from "@app/router";
 
   export let activeRoute: ProjectRoute;
   export let blob: Blob;
   export let getImage: (path: string) => Promise<Blob>;
   export let line: string | undefined = undefined;
 
-  $: lineNumber = line ? parseInt(line.substring(1)) : undefined;
-
+  const fileExtension = blob.path.split(".").pop() ?? "";
   const lastCommit = blob.info.lastCommit;
-  const lines = blob.binary ? 0 : (blob.content.match(/\n/g) || []).length;
-  const hasFinalNewline = blob.content.endsWith("\n");
-  const lineNumbers = Array(hasFinalNewline ? lines : lines + 1)
-    .fill(0)
-    .map((_, index) => (index + 1).toString());
   const parentDir = blob.path
     .match(/^.*\/|/)
     ?.values()
     .next().value;
+  let content: MaybeHighlighted = undefined;
+
+  onMount(async () => {
+    const output = await highlight(blob.content, fileExtension);
+    if (output) {
+      content = lineNumbersGutter(output);
+    }
+  });
 
-  // Waiting onMount, due to the line numbers still loading.
-  onMount(() => {
+  afterUpdate(() => {
     if (line) {
       scrollIntoView(line);
+
+      for (const item of document.getElementsByClassName("highlight")) {
+        item.classList.remove("highlight");
+      }
+
+      const element = document.getElementById(line);
+      if (element) {
+        element.classList.add("highlight");
+      }
```

Comment:
Could you help me understand why this is necessary?

Reply:
We don't have access to the line numbers, since they are part of the AST tree now.
So we have to add the `highlight` class somewhere, and we need to make sure any previous lines highlighted by the user are removed, else you can end up with multiple highlights.

We could eventually to make it clearer move the `classList.remove` function into an `beforeUpdate` hook

Reply:
Nice. Yeah, maybe also add a comment on what's going on.

Reply:
276110d0